
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Evaluating MAGIC with Synthetic Dataset &#8212; Academic Posts</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Markdown Files" href="markdown.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Academic Posts</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Welcome to the site!
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="markdown.html">
   Markdown Files
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Evaluating MAGIC with Synthetic Dataset
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://mybinder.org/v2/gh/saulvegasauceda/blog/main?urlpath=tree/docs/notebooks.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Launch on Binder"
>
  

<span class="headerbtn__icon-container">
  
    <img src="_static/images/logo_binder.svg">
  </span>
<span class="headerbtn__text-container">Binder</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/saulvegasauceda/blog"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/saulvegasauceda/blog/issues/new?title=Issue%20on%20page%20%2Fnotebooks.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="_sources/notebooks.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#imports">
   Imports:
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#functions">
     Functions:
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#generating-synthetic-dataset">
   1. Generating Synthetic Dataset
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#intro-to-merfish-10x-scrna-seq">
     1.1: Intro to MERFISH &amp; 10x scRNA-seq
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#generating-synthetic-cells-and-simulating-both-merfish-10x-scrna-seq">
     1.2: Generating synthetic cells and simulating both MERFISH &amp; 10x scRNA-seq
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#data">
     1.3: Data
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#peeking-at-the-data">
       Peeking at the data:
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#synthetic-data">
         Synthetic Data:
        </a>
       </li>
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#merfish-data">
         MERFISH data:
        </a>
       </li>
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#x-scrna-seq-data">
         10x scRNA-seq data:
        </a>
       </li>
      </ul>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#plotting-data">
       Plotting data:
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#imputation">
   2. Imputation
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#normalization">
     2.1: Normalization
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#running-magic">
     2.2: Running MAGIC
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#rescaling-imputed-counts">
     2.3: Rescaling Imputed Counts
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#results-after-imputation">
   3. Results after Imputation
  </a>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Evaluating MAGIC with Synthetic Dataset</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#imports">
   Imports:
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#functions">
     Functions:
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#generating-synthetic-dataset">
   1. Generating Synthetic Dataset
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#intro-to-merfish-10x-scrna-seq">
     1.1: Intro to MERFISH &amp; 10x scRNA-seq
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#generating-synthetic-cells-and-simulating-both-merfish-10x-scrna-seq">
     1.2: Generating synthetic cells and simulating both MERFISH &amp; 10x scRNA-seq
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#data">
     1.3: Data
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#peeking-at-the-data">
       Peeking at the data:
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#synthetic-data">
         Synthetic Data:
        </a>
       </li>
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#merfish-data">
         MERFISH data:
        </a>
       </li>
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#x-scrna-seq-data">
         10x scRNA-seq data:
        </a>
       </li>
      </ul>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#plotting-data">
       Plotting data:
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#imputation">
   2. Imputation
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#normalization">
     2.1: Normalization
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#running-magic">
     2.2: Running MAGIC
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#rescaling-imputed-counts">
     2.3: Rescaling Imputed Counts
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#results-after-imputation">
   3. Results after Imputation
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="evaluating-magic-with-synthetic-dataset">
<h1>Evaluating MAGIC with Synthetic Dataset<a class="headerlink" href="#evaluating-magic-with-synthetic-dataset" title="Permalink to this headline">#</a></h1>
<p><a class="reference external" href="https://github.com/KrishnaswamyLab/MAGIC">MAGIC</a> is an imputation tool developed at the <a class="reference external" href="https://www.broadinstitute.org">Broad Institute</a> for single cell RNA-sequencing (scNRA-seq) data. Imputation can potentially help improve downstream analysis in scRNA-seq data pipelines due to the potential of denoising the high dropout rates (errerroneously observed 0s) that currently obscure gene-gene relationships. Despite this promise, MAGIC has failed to receive universal support on whether it successfully recuperates lost information. This is an attempt to discover the effects of MAGIC on a synthetic dataset to determine whether its claims holds true for a simple test case.</p>
<section id="imports">
<h2>Imports:<a class="headerlink" href="#imports" title="Permalink to this headline">#</a></h2>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import magic
import pandas as pd 
import seaborn as sns
import scanpy as sc
import numpy as np
import matplotlib.pyplot as plt
import scanpy.external as sce
from numpy.random import seed
from numpy.random import normal
from random import sample
from random import choices
from scipy.stats import rv_discrete, rv_continuous
from scipy.stats import nbinom
</pre></div>
</div>
</div>
</div>
<section id="functions">
<h3>Functions:<a class="headerlink" href="#functions" title="Permalink to this headline">#</a></h3>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>seed(1738)
def create_synthetic_cells(rna_species_char, p, number_cells=100):
    &#39;&#39;&#39;
    Inputs
    --------
    p: the probability of a single success (parameter to scipy.stats.nbinom)
    rna_species_char: List of size N representing the profile of N RNA species
    number_cells: number of cells (rows) in synthetic dataset
    
    Sampling using a negative binomial distribution where n = # of median RNA
    
    Outputs
    --------
    Pandas dataframe of size (number_cells x N) where each row represents a cell
    sampled from rna_species_char.
    &#39;&#39;&#39;
    cell_generator = [nbinom(n, p) for n in rna_species_char]
    df = pd.DataFrame([dist.rvs(number_cells) for dist in cell_generator]).T
    df = df.set_axis([f&quot;Gene {g + 1}&quot; for g in range(len(df.columns))], axis=1, inplace=False)
    return df

sim_capture = lambda x, p: sum(choices([0, 1], weights=[1-p, p])[0] for _ in range(x))

def artificially_sample_cells(true_cells_df, capture_rate):
    &#39;&#39;&#39;
    Simulating Bernoulli sampling with synthetic dataset 
    where p = capture_rate (p := probability RNA is included in set)
    &#39;&#39;&#39;
    return true_cells_df.applymap(lambda x: sim_capture(x, p=capture_rate))

def ready_for_plot():
    merfish_repr[&quot;Model&quot;] = &quot;MERFISH&quot;
    tenx_repr[&quot;Model&quot;] = &quot;10X synthetic&quot;
    return pd.concat([merfish_repr, tenx_repr])

def rescale_data(data, target_data, rescale):
    &#39;&#39;&#39;
    Function from older version of MAGIC to rescale after imputation
    &#39;&#39;&#39;
    if rescale == 0:
        return data
    else:
        data[data &lt; 0] = 0
        M99 = np.percentile(target_data, rescale, axis=0)
        M100 = target_data.max(axis=0)
        indices = np.where(M99 == 0)[0]
        M99[indices] = M100[indices]
        M99_new = np.percentile(data, rescale, axis=0)
        M100_new = data.max(axis=0)
        indices = np.where(M99_new == 0)[0]
        M99_new[indices] = M100_new[indices]
        max_ratio = np.divide(M99, M99_new)
        data = np.multiply(data, np.tile(max_ratio,
                                      (target_data.shape[0], 1)))
    return data
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="generating-synthetic-dataset">
<h2>1. Generating Synthetic Dataset<a class="headerlink" href="#generating-synthetic-dataset" title="Permalink to this headline">#</a></h2>
<section id="intro-to-merfish-10x-scrna-seq">
<h3>1.1: Intro to MERFISH &amp; 10x scRNA-seq<a class="headerlink" href="#intro-to-merfish-10x-scrna-seq" title="Permalink to this headline">#</a></h3>
<p>Two popular methods for measuring RNA species post-mortem are: <a class="reference external" href="https://www.science.org/doi/10.1126/science.aaa6090">MERFISH</a> &amp; <a class="reference external" href="https://www.10xgenomics.com/blog/single-cell-rna-seq-an-introductory-overview-and-tools-for-getting-started">10x scRNA-seq</a>. Both these methods are able to identify which cells within a tissue sample a given mRNA molecule comes from. Besides MERFISH being a fluorescent approach and 10x scRNA-seq utilizes sequencing, they differ in their capture rates and number of observed RNA species. MERFISH is limited by the visual spectrum, thus, can only observe an order of ~100 RNA species but with a capture rate of ~90% (fraction of RNA transcripts observerd within the ~100 RNA species). Meanwhile, 10x scRNA-seq is not bound by the observed RNA species but observe mRNA with a <a class="reference external" href="https://kb.10xgenomics.com/hc/en-us/articles/360001539051-What-fraction-of-mRNA-transcripts-are-captured-per-cell-">much lower capture rate of 30-32%</a>.</p>
</section>
<section id="generating-synthetic-cells-and-simulating-both-merfish-10x-scrna-seq">
<h3>1.2: Generating synthetic cells and simulating both MERFISH &amp; 10x scRNA-seq<a class="headerlink" href="#generating-synthetic-cells-and-simulating-both-merfish-10x-scrna-seq" title="Permalink to this headline">#</a></h3>
<p>Sampling using a negative binomial distribution for the cells we will determine as the ground truth. Then, simulating both MERFISH &amp; 10x scRNA-seq as a Bernoulli process with the capture rate serving as the probability of a success.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>Rna_species_characteristic_numbers = [1, 3, 10, 30, 100, 300, 1000, 3000, 10000]

synth_cells = create_synthetic_cells(Rna_species_characteristic_numbers, p=0.5)
merfish_original_synth = artificially_sample_cells(synth_cells, 0.9)
tenx_original_synth = artificially_sample_cells(synth_cells, 0.3)

synth_cells[&quot;Model&quot;] = &quot;Ground Truth&quot;
merfish_original_synth[&quot;Model&quot;] = &quot;MERFISH&quot;
tenx_original_synth[&quot;Model&quot;] = &quot;10x scRNA-seq&quot;
</pre></div>
</div>
</div>
</div>
</section>
<section id="data">
<h3>1.3: Data<a class="headerlink" href="#data" title="Permalink to this headline">#</a></h3>
<section id="peeking-at-the-data">
<h4>Peeking at the data:<a class="headerlink" href="#peeking-at-the-data" title="Permalink to this headline">#</a></h4>
<section id="synthetic-data">
<h5>Synthetic Data:<a class="headerlink" href="#synthetic-data" title="Permalink to this headline">#</a></h5>
<div class="cell tag_hide-cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>synth_cells.head()
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Gene 1</th>
      <th>Gene 2</th>
      <th>Gene 3</th>
      <th>Gene 4</th>
      <th>Gene 5</th>
      <th>Gene 6</th>
      <th>Gene 7</th>
      <th>Gene 8</th>
      <th>Gene 9</th>
      <th>Model</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>1</td>
      <td>3</td>
      <td>8</td>
      <td>30</td>
      <td>121</td>
      <td>295</td>
      <td>940</td>
      <td>2987</td>
      <td>10138</td>
      <td>Ground Truth</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2</td>
      <td>3</td>
      <td>5</td>
      <td>21</td>
      <td>116</td>
      <td>296</td>
      <td>1002</td>
      <td>2970</td>
      <td>10366</td>
      <td>Ground Truth</td>
    </tr>
    <tr>
      <th>2</th>
      <td>8</td>
      <td>1</td>
      <td>17</td>
      <td>18</td>
      <td>133</td>
      <td>320</td>
      <td>1031</td>
      <td>3073</td>
      <td>9983</td>
      <td>Ground Truth</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0</td>
      <td>2</td>
      <td>8</td>
      <td>29</td>
      <td>118</td>
      <td>320</td>
      <td>983</td>
      <td>2861</td>
      <td>9927</td>
      <td>Ground Truth</td>
    </tr>
    <tr>
      <th>4</th>
      <td>7</td>
      <td>2</td>
      <td>8</td>
      <td>30</td>
      <td>85</td>
      <td>290</td>
      <td>1056</td>
      <td>2982</td>
      <td>9959</td>
      <td>Ground Truth</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
<section id="merfish-data">
<h5>MERFISH data:<a class="headerlink" href="#merfish-data" title="Permalink to this headline">#</a></h5>
<div class="cell tag_hide-cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>merfish_original_synth.head(5)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Gene 1</th>
      <th>Gene 2</th>
      <th>Gene 3</th>
      <th>Gene 4</th>
      <th>Gene 5</th>
      <th>Gene 6</th>
      <th>Gene 7</th>
      <th>Gene 8</th>
      <th>Gene 9</th>
      <th>Model</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>1</td>
      <td>3</td>
      <td>8</td>
      <td>28</td>
      <td>102</td>
      <td>269</td>
      <td>831</td>
      <td>2689</td>
      <td>9152</td>
      <td>MERFISH</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2</td>
      <td>3</td>
      <td>5</td>
      <td>19</td>
      <td>101</td>
      <td>267</td>
      <td>897</td>
      <td>2682</td>
      <td>9334</td>
      <td>MERFISH</td>
    </tr>
    <tr>
      <th>2</th>
      <td>5</td>
      <td>0</td>
      <td>15</td>
      <td>16</td>
      <td>117</td>
      <td>286</td>
      <td>925</td>
      <td>2742</td>
      <td>8898</td>
      <td>MERFISH</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0</td>
      <td>2</td>
      <td>8</td>
      <td>27</td>
      <td>107</td>
      <td>289</td>
      <td>876</td>
      <td>2588</td>
      <td>8936</td>
      <td>MERFISH</td>
    </tr>
    <tr>
      <th>4</th>
      <td>6</td>
      <td>2</td>
      <td>6</td>
      <td>27</td>
      <td>76</td>
      <td>255</td>
      <td>933</td>
      <td>2687</td>
      <td>9001</td>
      <td>MERFISH</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
<section id="x-scrna-seq-data">
<h5>10x scRNA-seq data:<a class="headerlink" href="#x-scrna-seq-data" title="Permalink to this headline">#</a></h5>
<div class="cell tag_hide-cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>tenx_original_synth.head(5)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Gene 1</th>
      <th>Gene 2</th>
      <th>Gene 3</th>
      <th>Gene 4</th>
      <th>Gene 5</th>
      <th>Gene 6</th>
      <th>Gene 7</th>
      <th>Gene 8</th>
      <th>Gene 9</th>
      <th>Model</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>0</td>
      <td>2</td>
      <td>7</td>
      <td>26</td>
      <td>87</td>
      <td>276</td>
      <td>905</td>
      <td>3042</td>
      <td>10x scRNA-seq</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0</td>
      <td>1</td>
      <td>1</td>
      <td>5</td>
      <td>31</td>
      <td>79</td>
      <td>313</td>
      <td>894</td>
      <td>3155</td>
      <td>10x scRNA-seq</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2</td>
      <td>0</td>
      <td>5</td>
      <td>7</td>
      <td>38</td>
      <td>96</td>
      <td>298</td>
      <td>901</td>
      <td>3078</td>
      <td>10x scRNA-seq</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0</td>
      <td>1</td>
      <td>3</td>
      <td>11</td>
      <td>31</td>
      <td>91</td>
      <td>295</td>
      <td>805</td>
      <td>2890</td>
      <td>10x scRNA-seq</td>
    </tr>
    <tr>
      <th>4</th>
      <td>1</td>
      <td>0</td>
      <td>2</td>
      <td>10</td>
      <td>21</td>
      <td>83</td>
      <td>338</td>
      <td>896</td>
      <td>2992</td>
      <td>10x scRNA-seq</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
</section>
<section id="plotting-data">
<h4>Plotting data:<a class="headerlink" href="#plotting-data" title="Permalink to this headline">#</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>before_imputation = pd.concat([synth_cells, merfish_original_synth, tenx_original_synth], ignore_index=True)
genes = before_imputation.columns[:-1]

for i in range(5): 
    plt.figure() #this creates a new figure on which your plot will appear
    sns.kdeplot(data=before_imputation, x=genes[i], y =genes[i+1], hue=&quot;Model&quot;, alpha=0.2, fill=True)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/notebooks_19_0.png" src="_images/notebooks_19_0.png" />
<img alt="_images/notebooks_19_1.png" src="_images/notebooks_19_1.png" />
<img alt="_images/notebooks_19_2.png" src="_images/notebooks_19_2.png" />
<img alt="_images/notebooks_19_3.png" src="_images/notebooks_19_3.png" />
<img alt="_images/notebooks_19_4.png" src="_images/notebooks_19_4.png" />
</div>
</div>
<p>From the plots, we can see that the MERFISH distributions more closely resemble that of the Ground Truth dataset. Thus, our goal with imputation is to transform the 10x scRNA-seq data to become more similar to the MERFISH data.</p>
</section>
</section>
</section>
<section id="imputation">
<h2>2. Imputation<a class="headerlink" href="#imputation" title="Permalink to this headline">#</a></h2>
<section id="normalization">
<h3>2.1: Normalization<a class="headerlink" href="#normalization" title="Permalink to this headline">#</a></h3>
<p>Normalizing the RNA counts as a processing step for MAGIC. Using the MERFISH as basis for normalized sum to compare downstream. Before MAGIC, the 10x scRNA-seq data is normalized by the median RNA count of the dataset, then the values are square rooted.</p>
<div class="cell tag_hide-output docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>merfish_original_synth = merfish_original_synth.drop(columns=[&quot;Model&quot;])
tenx_original_synth = tenx_original_synth.drop(columns=[&quot;Model&quot;])

# obtaining target sum for normalization
merfish_rna_counts = merfish_original_synth.sum(axis=1)
normalized_sum = int(merfish_rna_counts.median())

# Normalizing dataset
merfish_repr = sc.AnnData(merfish_original_synth)
tenx_repr = sc.AnnData(tenx_original_synth)

sc.pp.normalize_total(merfish_repr, target_sum=normalized_sum)
tenx_norm = tenx_repr.copy()
sc.pp.normalize_total(tenx_norm, target_sum=normalized_sum)

# Pre-processing step for MAGIC
sc.pp.normalize_total(tenx_repr)
tenx_raw = tenx_repr.to_df() # use later for rescaling
sc.pp.sqrt(tenx_repr)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/tmp/ipykernel_1989/1306700857.py:9: FutureWarning: X.dtype being converted to np.float32 from int64. In the next version of anndata (0.9) conversion will not be automatic. Pass dtype explicitly to avoid this warning. Pass `AnnData(X, dtype=X.dtype, ...)` to get the future behavour.
  merfish_repr = sc.AnnData(merfish_original_synth)
/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.
  warnings.warn(&quot;Transforming to str index.&quot;, ImplicitModificationWarning)
/tmp/ipykernel_1989/1306700857.py:10: FutureWarning: X.dtype being converted to np.float32 from int64. In the next version of anndata (0.9) conversion will not be automatic. Pass dtype explicitly to avoid this warning. Pass `AnnData(X, dtype=X.dtype, ...)` to get the future behavour.
  tenx_repr = sc.AnnData(tenx_original_synth)
</pre></div>
</div>
</div>
</div>
</section>
<section id="running-magic">
<h3>2.2: Running MAGIC<a class="headerlink" href="#running-magic" title="Permalink to this headline">#</a></h3>
<p>Running MAGIC with default parameters and normalizing the imputed counts back to the basis of the MERFISH median by squaring and normalizing sum.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%capture
sce.pp.magic(tenx_repr, name_list=&#39;all_genes&#39;, solver=&#39;exact&#39;, n_jobs=-1)
tenx_repr.X = tenx_repr.X**2
</pre></div>
</div>
</div>
</div>
</section>
<section id="rescaling-imputed-counts">
<h3>2.3: Rescaling Imputed Counts<a class="headerlink" href="#rescaling-imputed-counts" title="Permalink to this headline">#</a></h3>
<p>Duplicating the results of MAGIC to determine whether rescaling after imputation is still necessary. Using a vestigial function <code class="docutils literal notranslate"><span class="pre">rescale_data()</span></code> from the <a class="reference external" href="https://github.com/KrishnaswamyLab">Krishnaswamy Lab</a>.</p>
<div class="cell tag_hide-output docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>tenx_rescaled_norm = tenx_repr.copy()
imputed_rescaled_df = tenx_rescaled_norm.to_df()
imputed_rescaled_df = rescale_data(imputed_rescaled_df, tenx_raw, 99)
imputed_rescaled = sc.AnnData(imputed_rescaled_df)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/tmp/ipykernel_1989/2028333819.py:4: FutureWarning: X.dtype being converted to np.float32 from float64. In the next version of anndata (0.9) conversion will not be automatic. Pass dtype explicitly to avoid this warning. Pass `AnnData(X, dtype=X.dtype, ...)` to get the future behavour.
  imputed_rescaled = sc.AnnData(imputed_rescaled_df)
</pre></div>
</div>
</div>
</div>
<p>Here, we are re-normalizing the both the impute counts with and without rescaling to the previously used target sum.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pp.normalize_total(tenx_repr, target_sum=normalized_sum)
sc.pp.normalize_total(imputed_rescaled, target_sum=normalized_sum)

imputed_df = tenx_repr.to_df()
imputed_rescaled_df = imputed_rescaled.to_df()
merfish_df = merfish_repr.to_df()
tenx_norm_df = tenx_norm.to_df()

imputed_df[&quot;Model&quot;] = &quot;10x scRNA-seq Imputed&quot;
imputed_rescaled_df[&quot;Model&quot;] = &quot;10x Imputed Rescaled&quot;
tenx_norm_df[&quot;Model&quot;] = &quot;10x scRNA-seq&quot;
merfish_df[&quot;Model&quot;] = &quot;MERFISH Normalized&quot;
synth_cells[&quot;Model&quot;] = &quot;True cells&quot;

after_imputation = pd.concat([imputed_rescaled_df,merfish_df, tenx_norm_df, synth_cells], ignore_index=True)
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="results-after-imputation">
<h2>3. Results after Imputation<a class="headerlink" href="#results-after-imputation" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>for i in range(5): 
    plt.figure()
    sns.kdeplot(data=after_imputation, x=genes[i], y =genes[i+1], hue=&quot;Model&quot;, alpha=0.4, fill=True)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/notebooks_31_0.png" src="_images/notebooks_31_0.png" />
<img alt="_images/notebooks_31_1.png" src="_images/notebooks_31_1.png" />
<img alt="_images/notebooks_31_2.png" src="_images/notebooks_31_2.png" />
<img alt="_images/notebooks_31_3.png" src="_images/notebooks_31_3.png" />
<img alt="_images/notebooks_31_4.png" src="_images/notebooks_31_4.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>for i in range(5): 
    plt.figure()
    sns.scatterplot(data=after_imputation, x=genes[i], y =genes[i+1], hue=&quot;Model&quot;, alpha=0.2, s=40)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/notebooks_32_0.png" src="_images/notebooks_32_0.png" />
<img alt="_images/notebooks_32_1.png" src="_images/notebooks_32_1.png" />
<img alt="_images/notebooks_32_2.png" src="_images/notebooks_32_2.png" />
<img alt="_images/notebooks_32_3.png" src="_images/notebooks_32_3.png" />
<img alt="_images/notebooks_32_4.png" src="_images/notebooks_32_4.png" />
</div>
</div>
<p>Thinking of including a metric to evaluate the performance of imputation. I could also write another blog post about the metric.</p>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="markdown.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Markdown Files</p>
        </div>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Saul Vega Sauceda<br/>
  
      &copy; Copyright 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>